frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/services/ReverificationResultService.java

```diff
@@ -0,0 +1,201 @@
+                tokenResponse = TokenResponse.parse(tokenRequest.toHTTPRequest().send());
+                if (!tokenResponse.indicatesSuccess()) {
+                    HTTPResponse response = tokenResponse.toHTTPResponse();
+                    LOG.warn(
+                            "Unsuccessful %s response from IPV token endpoint on attempt %d: %s ",
+                            response.getStatusCode(), count, response.getContent());
+                }
+            } while (!tokenResponse.indicatesSuccess() && count < maxTries);
+
+            return tokenResponse;
+        } catch (IOException e) {
+            LOG.error("Error whilst sending TokenRequest", e);
+            throw new TokenRequestException("Error whilst sending TokenRequest", e);
+        } catch (com.nimbusds.oauth2.sdk.ParseException e) {
+            LOG.error("Error whilst parsing TokenResponse", e);
+            throw new TokenResponseException("Error whilst parsing TokenResponse", e);
+        }
+    }
+
+    public UserInfo sendIpvReverificationRequest(UserInfoRequest userInfoRequest)
+            throws UnsuccessfulReverificationResponseException {
+        try {
+            LOG.info("Sending IPV userinfo request");
+            int count = 0;
+            int maxTries = 2;
+            HTTPResponse response;
+            do {
+                if (count > 0) LOG.warn("Retrying IPV reverification request");
+                count++;
+                response = userInfoRequest.toHTTPRequest().send();
+                if (!response.indicatesSuccess()) {
+                    LOG.warn(
+                            "Unsuccessful %s response from IPV reverification endpoint on attempt %d: %s ",
+                            response.getStatusCode(), count, response.getContent());
+                }
+            } while (!response.indicatesSuccess() && count < maxTries);
+            if (!response.indicatesSuccess()) {
+                throw new UnsuccessfulReverificationResponseException(
+                        String.format(
+                                "Error %s when attempting to call IPV reverification endpoint: %s",
+                                response.getStatusCode(), response.getContent()));
+            }
+            LOG.info("Received successful reverification response");
+            return parseUserInfoFromResponse(response);
+        } catch (IOException e) {
+            throw new UnsuccessfulReverificationResponseException(
+                    "Error when attempting to call IPV reverification endpoint", e);
+        }
+    }
+
+    UserInfo parseUserInfoFromResponse(HTTPResponse response)
+            throws UnsuccessfulReverificationResponseException {
+        try {
+            String content = response.getContent();
+            if (content == null) {
+                throw new UnsuccessfulReverificationResponseException(
+                        "No content in HTTP response");
+            }
+            return UserInfo.parse(content);
```

Comment:
Just thinking about this... Is there any point deserialising this just to serialise it again? I don't think we read the result do we?

Reply:
if we do deserialise it, I think I would prefer a custom class so we get some validation of the shape. What do you think?

Reply:
I think we can just pass it on, no need to deserialise.